File: backend/api/memories.py
"""Memory management API endpoints."""

from fastapi import APIRouter, Depends, HTTPException, Request
from auth import AuthUser, get_current_user

from memory.store import MemoryStore
from memory.profile import ProfileManager

router = APIRouter(prefix="/api/memories", tags=["memories"])


@router.get("")
async def list_memories(user: AuthUser = Depends(get_current_user)):
    """List all non-deprecated memories."""
    store = MemoryStore(user.tenant_id, user.id)
    memories = await store.list_memories(include_deprecated=False)
    return {"count": len(memories), "memories": memories}


@router.get("/profile")
async def get_profile(user: AuthUser = Depends(get_current_user)):
    """Get the latest user profile."""
    pm = ProfileManager(user.tenant_id, user.id)
    profile = await pm.get_latest_profile()
    if profile is None:
        return {"profile": None, "message": "No profile generated yet"}
    return {"profile": profile}


@router.post("/profile/generate")
async def generate_profile(user: AuthUser = Depends(get_current_user)):
    """Trigger profile regeneration from current memories."""
    pm = ProfileManager(user.tenant_id, user.id)
    profile_data = await pm.generate_profile()
    if profile_data.get("total_memories", 0) == 0:
        return {"profile": None, "message": "No memories available to build profile"}
    profile_id = await pm.save_profile(profile_data, memory_ids=[])
    latest = await pm.get_latest_profile()
    return {"profile": latest, "message": "Profile regenerated successfully"}


@router.put("/{memory_id}")
async def update_memory(memory_id: str, request: Request,
                        user: AuthUser = Depends(get_current_user)):
    """Update a memory's content and/or deprecated flag."""
    body = await request.json()
    content = body.get("content")
    deprecated = body.get("deprecated")

    if content is None and deprecated is None:
        raise HTTPException(status_code=400, detail="No fields to update")

    store = MemoryStore(user.tenant_id, user.id)
    ok = await store.update_memory(memory_id, content=content, deprecated=deprecated)
    if not ok:
        raise HTTPException(status_code=404, detail="Memory not found")
    return {"ok": True}


@router.delete("/{memory_id}")
async def delete_memory(memory_id: str, user: AuthUser = Depends(get_current_user)):
    """Delete a single memory by ID."""
    store = MemoryStore(user.tenant_id, user.id)
    ok = await store.delete_memory(memory_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Memory not found")
    return {"ok": True}


@router.delete("")
async def clear_all_memories(user: AuthUser = Depends(get_current_user)):
    """Delete all memories."""
    store = MemoryStore(user.tenant_id, user.id)
    count = await store.clear_all()
    return {"ok": True, "deleted_count": count}

File: frontend/src/api/memories.ts
import { apiGet, apiPut, apiDelete, apiPost } from "./client";

export interface MemoryItem {
  id: string;
  content: string;
  memory_type: string;
  deprecated: boolean;
  created_at: string;
  updated_at: string;
}

export interface MemoriesResponse {
  count: number;
  memories: MemoryItem[];
}

export interface ProfileResponse {
  profile: {
    id: number;
    profile_data: Record<string, unknown>;
    version: number;
    generated_at: string;
  } | null;
  message?: string;
}

export function listMemories(): Promise<MemoriesResponse> {
  return apiGet("/api/memories");
}

export function updateMemory(id: string, body: { content?: string; deprecated?: boolean }): Promise<{ ok: boolean }> {
  return apiPut(`/api/memories/${id}`, body);
}

export function deleteMemory(id: string): Promise<{ ok: boolean }> {
  return apiDelete(`/api/memories/${id}`);
}

export function clearAllMemories(): Promise<{ ok: boolean; deleted_count: number }> {
  return apiDelete("/api/memories");
}

export function getProfile(): Promise<ProfileResponse> {
  return apiGet("/api/memories/profile");
}

export function generateProfile(): Promise<ProfileResponse> {
  return apiPost("/api/memories/profile/generate");
}
